transformer_engine/pytorch/ops/op.py

```diff
@@ -0,0 +1,428 @@
+# Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+
+"""Base classes for fusable operations."""
+
+from __future__ import annotations
+import abc
+from collections.abc import Iterable
+import dataclasses
+from typing import Any, Optional
+
+import torch
+
+import transformer_engine_torch as tex
+from transformer_engine.pytorch.fp8 import (
+    FP8GlobalStateManager,
+    get_default_fp8_recipe,
+)
+from ._common import canonicalize_device, is_float8_tensor
+
+
+@dataclasses.dataclass
+class OperationContext:
+    """State needed to apply an operation
+
+    Saves state from forward pass for use in backward pass.
+
+    """
+
+    # Tensors that have been saved from forward function
+    # Note: Available in the backward function, matching tensors from
+    # to_save.
+    saved_tensors: Optional[tuple[Optional[torch.Tensor], ...]] = None
+    # Tensors to save for backward function
+    # Note: Expected to be set in the forward function, either
+    # directly or with save_for_backward.
+    to_save: Optional[tuple[Optional[torch.Tensor], ...]] = None
+
+    # Corresponding range in pipeline's list of saved tensors
+    _saved_tensors_range: Optional[tuple[int, int]] = None
+
+    # Whether backward pass is required
+    _requires_grad: bool = False
+
+    def save_for_backward(self, *tensors: Optional[torch.Tensor]) -> None:
+        """Register tensors to be saved for the backward function
+
+        Expected to be called in the forward function.
+
+        """
+        self.to_save = tensors
+
+
+class FusableOperation(torch.nn.Module, metaclass=abc.ABCMeta):
```

Comment:
```suggestion
class FusibleOperation(torch.nn.Module, metaclass=abc.ABCMeta):
```

Reply:
Will require changes in all other instances

Reply:
My spelling instincts say "fusable" is more correct than "fusible", but I think you're right: https://www.merriam-webster.com/dictionary/fusible